src/python/pants/option/subsystem.py

```diff
@@ -80,9 +133,39 @@ def rule(cls) -> Rule:
             canonical_name=name,
         )
 
+    @classmethod
+    def construct_env_aware_rule(cls) -> Rule:
+        """Returns a `TaskRule` that will construct the target Subsystem.EnvironmentAware."""
+        # Global-level imports are conditional, we need to re-import here for runtime use
+        from pants.core.util_rules.environments import EnvironmentTarget
+        from pants.engine.rules import TaskRule
+
+        snake_scope = normalize_scope(cls.options_scope)
+        name = f"construct_env_aware_scope_{snake_scope}"
+
+        return TaskRule(
+            output_type=cls.EnvironmentAware,
+            input_selectors=(cls, EnvironmentTarget),
+            func=_construct_env_aware,
+            input_gets=(),
+            canonical_name=name,
+        )
+
     @memoized_classmethod
-    def rules(cls) -> Iterable[Rule]:
-        return [cls.rule()]
+    def rules(cls: Any) -> Iterable[Rule]:
+        from pants.core.util_rules.environments import add_option_fields_for
+        from pants.engine.rules import Rule
+
+        # nb. this needs to be memoized so that repeated calls to add these rules return
+        # exactly the same rule objects. As such, returning this generator directly
+        # won't work, because the iterator needs to be replayable.
```

Comment:
Are you sure? Also this isn't memoized iiuc

Reply:
The outer function is. I've amended the comment again.

And yes, we need to ensure the same rules are returned, as each invocation of `add_option_fields_for()` creates new `Field` objects, and things would go badly if such objects were re-registered